### PR TITLE
feat: upgrade support

### PR DESCRIPTION
**Pull request by ronag**
```js
client.upgrade({
  path: '/',
  method: 'GET',
  protocol: 'Websocket'
}, (err, data) => {
  const { headers, socket } = data
})
```

```js
client.connect({
  path: '/'
}, (err, data) => {
  const { statusCode, headers, socket } = data
})
```

### CODE REVIEW

**Comment by ronag** [comment]
@mcollina Is this something we want to have? Feedback on API?

**Comment by ronag** [comment]
@szmarczak @delvedor

**Comment by ronag** [comment]
Was thinking to do the same for connect and/or include connect here as well.

**Comment by mcollina** [comment]
Or maybe it should throw. What about calling request if the connection has been upgraded?

**Comment by ronag** [comment]
> What happens if there are request in flight? Or if there are a few pipelined? I don't think upgrade should throw.

No more requests should be dispatched after an upgrade until it has completed. Not implemented in this PR yet.

> I don't think upgrade should throw.

Not following. We already throw on unexpected upgrade?

**Comment by ronag** [comment]
Any feedback on the `protocol` argument? Seems ok?

**Comment by ronag** [comment]
> What about calling request if the connection has been upgraded?

What do you mean?

File: lib/client.js [hunk 1]
```diff
@@ -437,7 +446,26 @@ class Parser extends HTTPParser {
...
+      detachSocket(socket)
+
+      this.unconsume()
+      setImmediate(() => this.close())
+
+      this.socket = null
+
+      // this.currentBufferOffset is updated in nextTick
+      process.nextTick(() => {
+        const head = this.getCurrentBuffer().slice(this.currentBufferOffset)
```
Comment (ronag) [inline lib/client.js:458]: `socket.unshift()` should also work?

File: lib/client.js [hunk 2]
```diff
@@ -427,20 +441,36 @@ class Parser extends HTTPParser {
...
-      util.destroy(socket, new NotSupportedError('upgrade not supported'))
+      this.upgrade = true
+
+      detachSocket(socket)
+
+      this.unconsume()
+      setImmediate(() => this.close())
+
+      // this.currentBufferOffset is updated in nextTick
+      process.nextTick(() => {
```
Comment (ronag) [inline lib/client.js:459]: I don't like this nextTick  but can't think of a better way...

**Comment by ronag** [comment]
I think this is somewhat ready. Lacking some tests.

**Comment by ronag** [comment]
Notice a significant improvement over the node core API is that the `head` argument is not necessary.

**Comment by mcollina** [comment]
I mean what happens if we start an upgrade to websocket or http2 and there are request in the pipeline.

**Comment by ronag** [comment]
> I mean what happens if we start an upgrade to websocket or http2 and there are request in the pipeline.

Before or after the request?

If it's before, I would think the server doesn't start processing the upgrade until all other requests has completed. Though I'm not sure. Maybe we should change so that an upgrade is not dispatched while there are other requests at head of line.

If it's after they will stay queued until the upgrade is completed and a new socket has been assigned.

**Pull request merged**